src/Functions/extractKeyValuePairs.cpp

```diff
@@ -0,0 +1,271 @@
+#include "extractKeyValuePairs.h"
+
+#include <Columns/ColumnMap.h>
+#include <Columns/ColumnsNumber.h>
+#include <DataTypes/DataTypeMap.h>
+#include <DataTypes/DataTypeString.h>
+#include <Functions/ReplaceStringImpl.h>
+#include <Functions/keyvaluepair/src/KeyValuePairExtractorBuilder.h>
+#include <Common/assert_cast.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
+}
+
+/*
+ * In order to leverage DB::ReplaceStringImpl for a better performance, the default escaping processor needs
+ * to be overridden by a no-op escaping processor. DB::ReplaceStringImpl does in-place replacing and leverages the
+ * Volnitsky searcher.
+ * */
+struct NoOpEscapingProcessor : KeyValuePairEscapingProcessor<ExtractKeyValuePairs::EscapingProcessorOutput>
+{
+    explicit NoOpEscapingProcessor(char) { }
+
+    Response process(const ResponseViews & response_views) const override { return response_views; }
+};
+
+ExtractKeyValuePairs::ExtractKeyValuePairs()
+    : return_type(std::make_shared<DataTypeMap>(std::make_shared<DataTypeString>(), std::make_shared<DataTypeString>()))
+{
+}
+
+String ExtractKeyValuePairs::getName() const
+{
+    return name;
+}
+
+ColumnPtr ExtractKeyValuePairs::executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr &, size_t) const
+{
+    auto [data_column, escape_character, key_value_pair_delimiter, item_delimiter, enclosing_character, value_special_characters_allow_list]
+        = parseArguments(arguments);
+
+    auto extractor = getExtractor(
+        escape_character, key_value_pair_delimiter, item_delimiter, enclosing_character, value_special_characters_allow_list);
+
+    auto raw_columns = extract(extractor, data_column);
+
+    return escape(raw_columns);
+}
+
+bool ExtractKeyValuePairs::isVariadic() const
+{
+    return true;
+}
+
+bool ExtractKeyValuePairs::isSuitableForShortCircuitArgumentsExecution(const DataTypesWithConstInfo & /*arguments*/) const
+{
+    return false;
+}
+
+size_t ExtractKeyValuePairs::getNumberOfArguments() const
+{
+    return 0u;
+}
+
+DataTypePtr ExtractKeyValuePairs::getReturnTypeImpl(const DataTypes & /*arguments*/) const
+{
+    return return_type;
+}
+
+ExtractKeyValuePairs::ParsedArguments ExtractKeyValuePairs::parseArguments(const ColumnsWithTypeAndName & arguments)
+{
+    if (arguments.empty())
+    {
+        throw Exception(ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH, "Function " + String(name) + "requires at least one argument");
+    }
+
+    std::unordered_set<char> value_special_characters_allow_list;
+
+    auto data_column = arguments[0].column;
+
+    if (arguments.size() == 1u)
+    {
+        return ParsedArguments{data_column, {}, {}, {}, {}, value_special_characters_allow_list};
+    }
+
+    auto escape_character = arguments[1].column->getDataAt(0).toView().front();
```

Comment:
We may also check that all setting arguments (1-5) are constants.

Reply:
Can you give me an example of that? Not sure what that means

Reply:
I mean in sql arguments except first are typically constants `extractKeyValuePairs(some_column, '\\', ':', ',', '"')` and 
we don't support different chars for different rows (it would be confusing), so we can behave as `throwIf` do.

https://github.com/ClickHouse/ClickHouse/blob/c27f0d28a17e2aa71884a17c859a6c280597d8a6/src/Functions/throwIf.cpp#L69

```sql
select throwIf(0, 'message');
-- Ok

select throwIf(0, materialize('message'));
-- Code: 44. DB::Exception: Argument at index 1 for function throwIf must be constant. (ILLEGAL_COLUMN)

select throwIf(0, 'message' || toString(number)) from numbers(10);
-- Code: 44. DB::Exception: Argument at index 1 for function throwIf must be constant. (ILLEGAL_COLUMN)
```